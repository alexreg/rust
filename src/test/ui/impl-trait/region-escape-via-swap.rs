// compile-fail

// See https://github.com/rust-lang/rust/pull/57870#issuecomment-457333709 for more details.

trait Swap: Sized {
    fn swap(self, other: Self);
}

impl<T> Swap for &mut T {
    fn swap(self, other: Self) {
        std::mem::swap(self, other);
    }
}

// The hidden relifetimegion `'b` should not be allowed to escape this function, since it may not
// outlive '`a`, in which case we have a dangling reference.
fn hide<'a, 'b, T: 'static>(x: &'a mut &'b T) -> impl Swap + 'a {
//~^ lifetime mismatch [E0623]
    x
}

fn dangle() -> &'static [i32; 3] {
    let mut res = &[4, 5, 6];
    let x = [1, 2, 3];
    hide(&mut res).swap(hide(&mut &x));
    res
}

fn main() {}
