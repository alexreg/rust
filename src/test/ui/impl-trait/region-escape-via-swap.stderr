error[E0623]: lifetime mismatch
  --> $DIR/region-escape-via-swap.rs:17:50
   |
LL | fn hide<'a, 'b, T: 'static>(x: &'a mut &'b T) -> impl Swap + 'a {
   |                                        -----     ^^^^^^^^^^^^^^
   |                                        |         |
   |                                        |         ...but data from `x` is returned here
   |                                        this parameter and the return type are declared with different lifetimes...

error: aborting due to previous error

For more information about this error, try `rustc --explain E0623`.
